thetis/callback.py

```diff
@@ -481,3 +481,43 @@ def __call__(self):
             field_vals.append(np.reshape(self._evaluate_field(field_name), (ndetectors, -1)))
 
         return np.hstack(field_vals)
+
+
+class FunctionalCallback(DiagnosticCallback):
+    """Callback that evaluates a functional involving integrals in both time and space. This callback can also be used
+    to assemble time dependent objective functionals for adjoint simulations. Time integration is achieved using the
+    trapezium rule."""
+    variable_names = ['current integral', 'objective value']
+
+    def __init__(self, scalar_callback, solver_obj, **kwargs):
+        """
+        :arg scalar_callback: Python function that returns a list of values of an objective functional.
+        :arg solver_obj: Thetis solver object
+        :arg **kwargs: any additional keyword arguments, see DiagnosticCallback
+        """
+        kwargs.setdefault('export_to_hdf5', False)
+        kwargs.setdefault('append_to_log', False)
+        super(FunctionalCallback, self).__init__(solver_obj, **kwargs)
+        self.scalar_callback = scalar_callback      # Evaluate functional
+        self.objective_value = []   # Store functional value
```

Comment:
Here I'm wondering if using a python list is the best way of storing the values. It's easy to append to, but wastes some memory compared to numpy arrays.